### PR TITLE
Added room determinism through Elysium

### LINKED ISSUE #2: Design Fixed Room at Depth
Create an approach for fixing the room tiles to be consistent at each depth.

The definition of done for this story is to create follow up stories / some comprehensive documentation on how room generation will work.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1199160141707853/1199160141707897) by [Unito](https://www.unito.io/learn-more)
┆Complexity: 5
┆Priority: High

### PR DESCRIPTION
**Pull request by ellomenop**

### CODE REVIEW

File: RoomDeterminism/RoomDeterminism.lua [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+        story_depth = rng:Random(config.StoryMinDepth[biome], config.StoryMaxDepth[biome])
+      until #(placed_rooms[story_depth] or {}) < ((placed_rooms[story_depth] or {NumExits = MAX_EXITS[biome]}).NumExits or MAX_EXITS[biome]) -- Previous room has enough exits for this room
+            and #(placed_rooms[story_depth + 1] or {}) <= RoomSetData[biome][story_room_name].NumExits -- This room can support the number of needed exits for next room
+      placed_rooms[story_depth] = placed_rooms[story_depth] or {} -- Create if nil
+      table.insert(placed_rooms[story_depth], story_room_name)
+    end
+
+    -- For each depth, excluding the end shop, end boss and midbiome
+    for depth = BIOME_DEPTH[biome].Min, BIOME_DEPTH[biome].Max - 3 do
+      local minimum_exits = #(placed_rooms[depth + 1] or {1})
```
Comment (parasHadesMods) [inline RoomDeterminism/RoomDeterminism.lua:163]: Usually the shop can only appear after 2-exit rooms.

If the shop is at depth + 1, maybe bump minimum_exits to 2.

File: RoomDeterminism/RoomDeterminism.lua [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+    eligible_room_data[intro_room_name] = RoomSetData[biome][intro_room_name]
+    return eligible_room_data
+  end
+
+  -- Look through all room data to find the eligible ones
+  for roomName, roomData in pairs(RoomSetData[biome]) do
+    -- Filter down to only combat chambers that aren't debug rooms
+    if string.find(roomName, "Combat") and not roomData.DebugOnly then
+      -- Check all other conditions (e.g. biome depth and num exits)
+      if roomData.NumExits >= minimum_exits
```
Comment (parasHadesMods) [inline RoomDeterminism/RoomDeterminism.lua:210]: There are several rooms where roomData.NumExits is set to 1 but the room actually has 2 exits:
- B_Combat10
- C_MiniBoss02
- C_Reprieve01

This generally has the effect of preventing the midshop from appearing after these rooms. However, for answering "can this room lead to all the rooms in the next level", these rooms should be considered to have 2 exits.

File: RoomDeterminism/RoomDeterminism.lua [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+      intro_room_name = "RoomOpening"
+    end
+    eligible_room_data[intro_room_name] = RoomSetData[biome][intro_room_name]
+    return eligible_room_data
+  end
+
+  -- Look through all room data to find the eligible ones
+  for roomName, roomData in pairs(RoomSetData[biome]) do
+    -- Filter down to only combat chambers that aren't debug rooms
+    if string.find(roomName, "Combat") and not roomData.DebugOnly then
```
Comment (parasHadesMods) [inline RoomDeterminism/RoomDeterminism.lua:208]: This will exclude the Athena Room (RoomSimple01).

File: RoomDeterminism/RoomDeterminism.lua [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+      -- Select room from eligible rooms
+      local selected_room_name = GetRandomKey(eligible_room_data)
+      RoomOverridesByDepth[depth] = RoomOverridesByDepth[depth] or {} -- Create if nil
+      RoomOverridesByDepth[depth].Room = selected_room_name
+      SelectedRoomsCache[selected_room_name] = true
+      DebugPrint({Text = depth .. ": Selected " .. selected_room_name})
+
+      -- Generate a table of options based on number of exits of selected room
+      local exit_options = {}
+      for i = 1, eligible_room_data[selected_room_name].NumExits do
```
Comment (parasHadesMods) [inline RoomDeterminism/RoomDeterminism.lua:175]: See the note below re: NumExits, it's not always correct.

File: RoomDeterminism/RoomDeterminism.lua [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+          local exit = RemoveRandomValue(exit_options)
+          RoomOverridesByDepth[depth]["Exit" .. exit] = placed_room
+          DebugPrint({Text = depth .. ": Placed " .. placed_room .. " at " .. "Exit" .. exit})
+        end
+      end
+    end
+  end
+end
+
+-- TODO: How to handle RequiredFalseRooms / Combat08A in tartarus?
```
Comment (parasHadesMods) [inline RoomDeterminism/RoomDeterminism.lua:191]: If you pass in the list of selected rooms at depth n-1, then we can disallow Combat08A if any of them are in the list.

**Review by parasHadesMods — APPROVED** [review]

**Comment by Museus** [comment]
This was added in another commit (be129f18ab0cf4f0d9da0b0c41133b7125e8b2b9), made issues for comments on this PR

**Pull request closed without merging**